### ISSUE TITLE
Random numbers section

### ISSUE THREAD

**Issue by cubbimew** [isocpp/CppCoreGuidelines#869]
Suggestion for the [SL: The Standard Library](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#S-stdlib): could it discuss best practices w.r.t pseudo-random number generation? Something like
* avoid rand() (explicitly discouraged by the standard itself, also CERT rule 12.1 MSC50-CPP), use the `<random>` facilities instead
* seed random number generators appropriately (CERT rule 12.2 MSC51-CPP bans default-init and use of current time, and endorses std::random_device, famously non-random in mingw. I am not sure it's worth getting into the whole discussion of entropy, PRNG state size, std::seed_seq, etc, but a simple guideline could do good if it can be formulated)
* do not use modulo operator to resize a uniform distribution (it makes it non-uniform; use [std::uniform_int_distribution](http://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution))
* use [std::bernoulli_distribution](http://en.cppreference.com/w/cpp/numeric/random/bernoulli_distribution) to generate random bits (because a separate call to RNG for every bit, as some people do, is a waste) - maybe not, that's probably not detectable
* use a special-purpose library when using random numbers for cryptographic purposes... okay that's not detectable either.

**Comment by MikeGitb** [comment]
Sounds good.

Isn't there also some paper about simplifying the usage of <random> in the pipeline? I believe the main obstacle on the wide adoption of the `<random>` is the extreme verbosity of the "right" way to work with random numbers. In particular, when you "just want a random number" `rand()%d +a` is just so much simpler. I admit that in many cases where I don't have nice wrapper function handy I'll just write

    int my_random_number = std::uniform_int_distribution<int>{a,b}(std::random_device{});

**Comment by jwakely** [comment]
http://en.cppreference.com/w/cpp/experimental/randint

**Comment by cubbimew** [comment]
>http://en.cppreference.com/w/cpp/experimental/randint

huh, works in gcc.

**Comment by MikeGitb** [comment]
Yes, that's it. I forgot it was part of te TS. I think this should also be part of that section (e.g. a note about what to do if you just want a random number).

**Comment by AndrewPardoe** [comment]
Per our sync, perhaps we should have a short item that points to Walter's committee paper from about two years back.

**Comment by jwakely** [comment]
I think http://open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3847.pdf is the paper, and http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0346r0.pdf makes another key point: you should not use a URBG directly to get a random number, you should only treat them as source of random bits and use a distribution to shape that into a random number.

**Comment by DemiMarie** [comment]
I think the C++ standard should provide a random number generator specifically for cryptographic purposes.

**Comment by jwakely** [comment]
@DemiMarie that would be nice, but somebody has to propose it or it won't happen, and this isn't the place to propose it. See https://isocpp.org/std/submit-a-proposal for how to submit a proposal for standardization.
